src/compose/ports.ts

```diff
@@ -0,0 +1,206 @@
+import TypedError = require('typed-error');
+
+// Adapted from https://github.com/docker/docker-py/blob/master/docker/utils/ports.py#L3
+const PORTS_REGEX =
+	/^(?:(?:([a-fA-F\d.:]+):)?([\d]*)(?:-([\d]+))?:)?([\d]+)(?:-([\d]+))?(?:\/(udp|tcp))?$/;
+
+// A regex to extract the protocol and internal port of the incoming Docker options
+const DOCKER_OPTS_PORTS_REGEX =  /(\d+)(?:\/?([a-z]+))?/i;
+
+export class InvalidPortDefinition extends TypedError { }
+
+export interface PortBindings {
+	[key: string]: Array<{ HostIp: string, HostPort: string }>;
+}
+
+export interface DockerPortOptions {
+	exposedPorts: Map<string, {}>;
+	portBindings: PortBindings;
+}
+
+interface PortRange {
+	internalStart: number;
+	internalEnd: number;
+	externalStart: number;
+	externalEnd: number;
+	protocol: string;
+	host: string;
+}
+
+export class PortMap {
+
+	private internalStart: number;
+	private internalEnd: number;
+	private externalStart: number;
+	private externalEnd: number;
+	private protocol: string;
+	private host: string;
+
+	public constructor(portStrOrObj: string | PortRange) {
+		if (_.isString(portStrOrObj)) {
+			this.parsePortString(portStrOrObj);
+		} else {
+			this.internalStart = portStrOrObj.internalStart;
+			this.internalEnd = portStrOrObj.internalEnd;
+			this.externalStart = portStrOrObj.externalStart;
+			this.externalEnd = portStrOrObj.externalEnd;
+			this.protocol = portStrOrObj.protocol;
+			this.host = portStrOrObj.host;
+		}
+	}
+
+	public toDockerOpts(): DockerPortOptions {
+		const internalRange = this.generatePortRange(this.internalStart, this.internalEnd);
+		const externalRange = this.generatePortRange(this.externalStart, this.externalEnd);
+
+		const exposed: { [key: string]: {} } = {};
+		const portBindings: PortBindings = {};
+
+		_.each(_.zip(internalRange, externalRange), ([internal, external]) => {
+			exposed[`${internal}/${this.protocol}`] = {};
+
+			portBindings[`${internal}/${this.protocol}`] = [
+				{ HostIp: this.host, HostPort: external!.toString() },
+			];
+		});
+
+		return {
+			exposedPorts: exposed,
+			portBindings,
+		};
+	}
+
+	/**
+	 * fromDockerOpts
+	 *
+	 * We need to be able to compare target and running containers, but
+	 * running containers store their ports in a separate format (see the
+	 * output of toDockerOpts).
+	 *
+	 * This function takes the output of toDockerOpts (or more accurately,
+	 * the output of the docker daemon when probed about a given container)
+	 * and produces a list of PortMap objects, which can then be compared.
+	 *
+	 */
+	public static fromDockerOpts(
+		portBindings: PortBindings,
+	): PortMap[] {
+
+		// Create a list of portBindings, rather than the map (which we can't
+		// order)
+		const portMaps = _.map(portBindings, (hostObj, internalStr) => {
+
+			const match = internalStr.match(DOCKER_OPTS_PORTS_REGEX);
+			if (match == null) {
+				throw new Error(`Could not parse docker port output: ${internalStr}`);
+			}
+			const internal = parseInt(match![1], 10);
```

Comment:
The `!` shouldn't be needed here as we already verified to the compiler that `match` cannot be null